textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -33,9 +33,138 @@
 	
 	GramBooleanQuery(QueryOp operator, int gramLength) {
 		this.operator = operator;
-		operandList = new ArrayList<String>();
-		subQueryList = new ArrayList<GramBooleanQuery>();
-		this.gramLength = gramLength;
+		operandSet = new HashSet<String>();
+		subQuerySet = new HashSet<GramBooleanQuery>();
+		GramBooleanQuery.gramLength = gramLength;
+	}
+	
+	
+	/**
+	 * This method takes a list of strings and adds them to the query tree. <br>
+	 * For example, if the list is {abcd, wxyz}, then: <br>
+	 * trigrams({abcd, wxyz}) = trigrams(abcd) OR trigrams(wxyz) <br>
+	 * OR operator is assumed for a list of strings. <br>
+	 * @param list, a list of strings to be added into query.
+	 */
+	void add(List<String> list) {
+		addOrNode(list);
+	}
+	
+	private void addOrNode(List<String> literalList) {
+		if (TranslatorUtils.minLenOfString(literalList) < 3) {
+			return;
+		}
+		if (literalList.size() == 0) {
+			return;
+		} else if (literalList.size() == 1) {
+			this.addAndNode(literalList.get(0));
+		} else {
+			GramBooleanQuery query = new GramBooleanQuery(QueryOp.OR);
+			for (String literal : literalList) {
+				query.addAndNode(literal);
+			}
+			this.subQuerySet.add(query);
+		}
+	}
+	
+	/**
+	 * This method takes a single string and adds it to the query tree. <br>
+	 * The string is converted to multiple n-grams with an AND operator. <br>
+	 * For example: if the string is abcd, then: <br>
+	 * trigrams(abcd) = abc AND bcd <br>
+	 * AND operator is assumed for a single string. <br>
+	 * @param literal
+	 */
+	private void addAndNode(String literal) {
+		if (literal.length() < gramLength) {
+			return;
+		} else if (literal.length() == gramLength) {
+			this.operandSet.add(literal);
+		} else {
+			if (this.operator == QueryOp.AND) {
+				this.operandSet.addAll(literalToNGram(literal));
+			} else {
+				GramBooleanQuery query = new GramBooleanQuery(GramBooleanQuery.QueryOp.AND);
+				query.operandSet.addAll(literalToNGram(literal));
+				this.subQuerySet.add(query);
+			}
+		}
+	}
+	
+	/**
+	 * This function builds a list of N-Grams that a given literal contains. <br>
+	 * If the length of the literal is smaller than N, it returns an empty list. <br>
+	 * For example, for literal "textdb", its tri-gram list should be ["tex", "ext", "xtd", "tdb"]
+	 * @param literal
+	 * @return
+	 */
+	private List<String> literalToNGram(String literal) {
+		ArrayList<String> nGrams = new ArrayList<>();
+		if (literal.length() >= gramLength) {
+			for (int i = 0; i <= literal.length()-gramLength; ++i) {
+				nGrams.add(literal.substring(i, i+gramLength));
+			}
+		}
+		return nGrams;
+	}
+	
+	/**
+	 * This function "AND" two query trees together. <br>
+	 * It also performs simple simplifications. <br>
+	 * TODO: add more logic for more complicated and effecitve simplifications
+	 * @param that GramBooleanQuery
+	 * @return
+	 */
+	GramBooleanQuery and (GramBooleanQuery that) {
```

Comment:
Maybe we could name it `conjunct`? This function is computing the conjunction of `this` with `that`.